File: ProjectEuler/051-100/S75.cs
using System.Collections.Generic;
using System.Linq;

namespace ProjectEuler
{
    public class S75 : ISolution
    {
        private static readonly long L = 1500000;

        public string GetAnswer()
        {
            var countMapping = new Dictionary<long, int>();
            for (long a = 1; a < L / 4; a += 2)
            {
                for (long b = a + 2; a * b < L / 2; b += 2)
                {
                    if (Utils.GetGcd(a, b) == 1)
                    {
                        for (long k = 1; k * a * b < L / 2; k++)
                        {
                            long x = k * a * b;
                            long y = k * (b * b - a * a) / 2;
                            long z = k * (b * b + a * a) / 2;
                            long c = x + y + z;
                            if (c <= L)
                            {
                                if (countMapping.ContainsKey(c))
                                {
                                    countMapping[c]++;
                                }
                                else
                                {
                                    countMapping[c] = 1;
                                }
                            }
                        }
                    }
                }
            }

            return countMapping.Count(p => p.Value == 1).ToString();
        }
    }
}

File: src/051-100/S75.cc
#include "S75.h"
#include "../Utils.h"
#include <map>
#include <vector>

using namespace std;

namespace ProjectEuler
{
std::string S75::GetAnswer()
{
	int limit = 1500000;
	std::vector<int> counts(limit + 1, 0);

	// L = 2km(m+n)
	// m > n > 0
	// gcd(m, n) == 1
	// m, n not both odd (one even, one odd)

	int64_t m_limit = sqrt(limit / 2);

	for (int64_t m = 2; m <= m_limit; m++)
	{
		for (int64_t n = 1; n < m; n++)
		{
			if ((m + n) % 2 == 1 && Utils::GetGcd(m, n) == 1)
			{
				int64_t L = 2 * m * (m + n);
				int64_t k = 1;
				while (k * L <= limit)
				{
					counts[k * L]++;
					k++;
				}
			}
		}
	}

	int result = 0;
	for (int c : counts)
	{
		if (c == 1)
		{
			result++;
		}
	}

	return std::to_string(result);
}
} // namespace ProjectEuler
